sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/BroadcastExchangeExec.scala

```diff
@@ -61,23 +61,43 @@ trait BroadcastExchangeLike extends Exchange {
    */
   def relationFuture: Future[broadcast.Broadcast[Any]]
 
+  @transient
+  private lazy val promise = Promise[Unit]()
+
+  @transient
+  private lazy val scalaFuture: scala.concurrent.Future[Unit] = promise.future
+
+  @transient
+  private lazy val triggerFuture: Future[Any] = {
+    SQLExecution.withThreadLocalCaptured(session, BroadcastExchangeExec.executionContext) {
+      // Trigger broadcast preparation which can involve expensive operations like waiting on
+      // subqueries and file listing.
+      executeQuery(null)
```

Comment:
shall we do tryFailure if fail to executeQuery like shuffle ?
```
try {
  executeQuery(null)
  promise.trySuccess(())
} catch {
  case e: Throwable =>
    promise.tryFailure(e)
    throw e
}
```

Reply:
good point.